[Mobile Payments] Add a “Learn more” action to the connecting dialog

iamgabrielma opened the pull request:
Closes: #7518
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
As part of pdfdoF-1cz-p2, this PR adds a “Learn more” action to the Card Reader connecting dialog. This appears as a tappable "Learn More about In-Person Payments" text section just below the `Cancel` button when the app is scanning for readers. 

This is also tracked to Woo Events under `card_present_connection_learn_more_tapped`

### Changes
- Adds a new case to `PaymentsModalActionsMode`, which describes a new Modal that has a secondary button, and attributed text as its title.
- Adds a protocol extension to `CardPresentPaymentsModalViewModel` so we can provide with a nil default value to all other modals that conform to this protocol but do not use the new `auxiliaryAttributedButtonTitle`
- Adds logic to `configureAuxiliaryButton()` based on the view model that will be used to fill the Modal. We start by clearing any previous data via `clearAuxiliaryButton()`, then populate the contents of the Modal.
- Tapping the button launches a WebView to the "Getting started with IPP" website, as well as track the tap via the existing `.cardPresentOnboardingLearnMoreTapped`

### Testing instructions
1. If testing only with the simulator, enable -simulate-stripe-card-reader at Product > Scheme > Edit Scheme
2. Disable your Card Reader at HubMenu > Payments > Manager Card Readers > Disconnect
3. Open an Order that is pending payment > Tap Collect Payment
4. See the app Scanning for reader
5. Tapping on the text will redirect you to https://docs.woocommerce.com/document/getting-started-with-in-person-payments-with-woocommerce-payments/

As the flow doesn't require user interaction to change modals, for easier checking/testing you can override `viewDidAppear()` in `CardPresentPaymentsModalViewController` and add a breakpoint there ( https://github.com/woocommerce/woocommerce-ios/pull/7545/commits/c8dde2625d2108725195f1cb18cec2a2ebc305d9)
```
override func viewDidAppear(_ animated: Bool) {}
```
### Screenshots
<img width= 400 src="https://user-images.githubusercontent.com/3812076/186670917-d6591345-ccbc-4edb-b3dd-64ea814e79e6.png">

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

iamgabrielma committed:
Add “learn more” text to ScanningForReader modal

iamgabrielma committed:
Add WebViewHelper call to didTapAuxiliaryButton

iamgabrielma committed:
Add icon to learn more message. Add font style

iamgabrielma committed:
Extract attributtedString button logic to its viewModel

iamgabrielma committed:
Add multiple attributes to attributedString

iamgabrielma committed:
Style “learn more” button title attributedString

iamgabrielma committed:
Add tracking cardPresentOnboardingLearnMoreTapped

iamgabrielma committed:
Remove debugging viewDidAppear() breakpoint

iamgabrielma committed:
Add clearButton method. Add spacing to icon

joshheald commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalScanningForReader.swift:
```diff
@@ -21,6 +21,18 @@ final class CardPresentModalScanningForReader: CardPresentPaymentsModalViewModel
+        let linkAttributedString = NSMutableAttributedString(string: Localization.learnMoreLink, attributes: moreLinkAttributes)
+        let moreAttributedString = NSMutableAttributedString(string: Localization.learnMoreText, attributes: moreTextAttributes)
+        linkAttributedString.append(moreAttributedString)
+        return linkAttributedString
```
Unfortunately, concatenating the strings doesn't allow good translations in all languages. 

The approach to use is a format string with a placeholder for the link, and then range-replacement of the link text with formatted link text. See the example in InPersonPaymentsLearnMore for how that can be done.

iamgabrielma replied:
Thanks! I had problems making the link appear following this approach, and was because I didn't add the `%1$@` placeholder to the `NSLocalizedString`. Changed here: 73438f7dbf37b2887841d3b217ff8688c635b7c5

Edit: Forgot to mention, it would seem we can create some extension from this NSMutableAttributedString, as we tend to use the combination of "learn more" + text pretty commonly across the app. I'll check that outside of this PR.

joshheald commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift:
```diff
@@ -277,14 +279,29 @@ private extension CardPresentPaymentsModalViewController {
 
-        auxiliaryButton.setTitleWithoutAnimation(viewModel.auxiliaryButtonTitle, for: .normal)
         auxiliaryButton.isHidden = false
+        auxiliaryButton.setTitle(viewModel.auxiliaryButtonTitle, for: .normal)
```
Perhaps this should stay as `withoutAnimation`?

iamgabrielma replied:
Oh! Definitely, changed here: ac03b136ef4999cf6f485281dcf00ef0341e1806

joshheald commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift:
```diff
@@ -277,14 +279,29 @@ private extension CardPresentPaymentsModalViewController {
+        auxiliaryButton.setTitle(viewModel.auxiliaryButtonTitle, for: .normal)
         auxiliaryButton.accessibilityIdentifier = Accessibility.auxiliaryButton
+
+        if viewModel.actionsMode == .secondaryActionAndAttributedAuxiliaryButton {
+            auxiliaryButton.setImage(.infoOutlineImage, for: .normal)
+            auxiliaryButton.setAttributedTitle(viewModel.auxiliaryAttributedButtonTitle, for: .normal)
+            auxiliaryButton.distributeTitleAndImage(spacing: 12.0)
+        }
```
I don't think there's anything about the `actionsMode = .secondaryActionAndAttributedAuxiliaryButton` that logically says you should set an image and an attributed title here.

I'd suggest that you set them to the value in the viewModel (and put the image on the viewModel) regardless of whether they're nil or populated. You could put the `distributeTitleAndImage` call behind a check for `if viewModel.auxiliaryButtonImage == nil` if that would help.

iamgabrielma replied:
My main concern when I initially tried to make these changes was that the viewModel used here is not specifically our `CardPresentModalScanningForReader`, but the `CardPresentPaymentsModalViewModel` protocol, so I would need a way to filter what content I'm loading. 

I could use something like `viewModel as? CardPresentModalScanningForReader != nil` for the check, but this feels wrong as would compile but may not result on what we want on runtime, also would not resolve the issue of `auxiliaryButtonImage` not being part of the viewModel protocol, so I would need to add another default implementation for `auxiliaryButtonImage` to the private extension, as I already had to do with `auxiliaryAttributedButtonTitle`.

Happy to go through this part on another session, I may be miss-understanding how to filter down this properly 👍

joshheald commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift:
```diff
@@ -277,14 +279,29 @@ private extension CardPresentPaymentsModalViewController {
+    }
+
+    func clearAuxiliaryButton() {
+        auxiliaryButton.setImage(nil, for: .normal)
+        auxiliaryButton.setAttributedTitle(nil, for: .normal)
+        auxiliaryButton.setTitle(nil, for: .normal)
+        auxiliaryButton.accessibilityIdentifier = nil
```
This method _may_ become unnecessary as well, if you do the above... because things that should be nil will be set to nil based on the viewModel property.

joshheald reviewed (approved):
There are a few things here which we chatted about in pairing, it's probably helpful to have them written down... this change would be a bit better if you make these tweaks to it.

I think it would be best to make sure the image and both titles are optional properties from the viewModel, then set the button image/title based on the viewModel for every time the `configureAuxiliaryButton` method is called. That way you don't need to care if there's nil or something in those properties, you'll correctly set the button to nil when the property is nil.

Totally understand why you didn't go the enum route we tried here but didn't want this to get lost, we probably tried to cover too much in the one session. Give me a shout if you want to go over it again.

iamgabrielma committed:
Remove animation when changing button title

iamgabrielma committed:
Change attributtedString formatting for better L10n

iamgabrielma committed:
Change logic when setting auxiliary content

Rather than checking the viewModel’s actionsMode to populate the modal, we directly set the values from the viewModel, regardless if these are populated or not.

iamgabrielma commented:
Thanks for another great pairing session! I made the changes on 94a97e7105b5b288e15839f68d8fc0306c88fc28 

To avoid the UI flickering when loading new content, we only need to populate content inside `performWithoutAnimation() `and keep `view.layoutIfNeeded()` 👍

joshheald commented on WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalScanningForReader.swift:
```diff
@@ -21,16 +21,27 @@ final class CardPresentModalScanningForReader: CardPresentPaymentsModalViewModel
-        let moreAttributedString = NSMutableAttributedString(string: Localization.learnMoreText, attributes: moreTextAttributes)
-        linkAttributedString.append(moreAttributedString)
-        return linkAttributedString
+        let result = NSMutableAttributedString(
+            string: .localizedStringWithFormat(
+                Localization.learnMoreText,
+                Localization.learnMoreLink
+            ),
+            attributes: [.foregroundColor: UIColor.text]
+        )
+        result.replaceFirstOccurrence(
+            of: Localization.learnMoreLink,
+            with: NSAttributedString(
+                string: Localization.learnMoreLink,
+                attributes: [
+                    .foregroundColor: UIColor.accent,
+                    .underlineStyle: NSUnderlineStyle.single.rawValue
+                ]
+            ))
+        result.addAttribute(.font, value: UIFont.footnote, range: NSRange(location: 0, length: result.length))
+        return result
```
In my next PR there will be a `LearnMoreViewModel`, which you may want to use to produce this string. It's just extracted from the SwiftUI view. Can be in a later PR, it would be best to get this merged 😊

iamgabrielma replied:
Nice! I'll keep an eye on it :D

joshheald commented on WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsModalViewController.swift:
```diff
@@ -158,7 +158,7 @@ private extension CardPresentPaymentsModalViewController {
 
     func styleAuxiliaryButton() {
-        if viewModel.actionsMode != .secondaryActionAndAttributedAuxiliaryButton {
+        if viewModel.actionsMode != .secondaryActionAndAuxiliaryButton {
```
I didn't notice this one before but again it feels like a misuse of the `actionsMode`. Another view may in future want a secondary action and a link-styled auxiliary button.

Perhaps a better property to check would be `viewModel.attributedAuxiliaryButtonTitle` because it seems fairly safe to assume that we don't want to re-style an attributed string as a link?

joshheald reviewed (approved):
Works well on WCPay sites ... *foreshadowing*

On sites which use the Stripe Gateway, it still links to the WCPay docs. I did the same thing in the CoD enable screen 😬 #7598 

You can grab the correct URL for Stripe from the WooConstants as well, but feel free to do it in a separate PR: it's a minor issue really.

I've made a couple of suggestions. As always, let me know if you want to go through any of them. 

Sorry there's been a bit of back and forth on this one, I do think it's best to merge now, so please feel free to tackle anything in a follow up, or not at all if you think that there's more impactful work you need to get on to.
